### ISSUE TITLE
[4.0.0-4c.0] Provide full Swiper API

### ISSUE THREAD

**Issue by JCKodel** [ionic-team/ionic#16887]
# Feature Request

**Ionic version:**

[x] **4.0.0-rc.0**

**Describe the Feature Request**

`<ion-slides>` is based on http://idangero.us/swiper/api, but do not implement all features (swipe effects other than slides, parallax, etc.). Mixing Swiper with Ionic lead to problems (and duplicate code).

Giving access to the entire Swiper component power would be awesome.

**Describe Preferred Solution**

 As Swiper is modular, would be nice to have some way to import features on `<ion-slides>`, for example:

```
import SwiperParallax from "somewhere";

IonSlidesHTMLElement.modules = [SwiperParallax];
```

Or, at least, give access to the complete Swiper API (i.e.: make effect, parallax, etc. to work)

**Describe Alternatives**

No alternatives found, as importing Swiper seems to have side-effects on `<ion-slides>`.

**Additional Context**

Link to the original component: http://idangero.us/swiper/api
Link to the documentation: https://beta.ionicframework.com/docs/api/slides (it says it adopted the above library, but I think the build process have removed some features, as they can be optional)

**Comment by NG-tylerholden** [comment]
Many broken options and features with the swiper.

**Comment by CFT-Chris** [comment]
PR #16616 is in the queue to address the missing effect modules.

I think the issue with dynamic importing of modules is that Ionic 4 is meant to be a static, standalone library. There is no intermediate step to dynamically build and pack a new core/dist/collection/components/slides/swiper/swiper.bundle.js for all the modules you want included.

If all modules are included in swiper.bundle.js, then ``<ion-slides>`` becomes as bloated (and buggy) as the entire Swiper library. Also, if the entire Swiper library is included, the other modules may introduce elements and behavior not yet styled consistently with the platform-specific themes in Ionic.

If you simply import the full swiper library and just not use the ``<ion-slides>`` component, does that not work for you? The following worked for me:

``npm i swiper`` 

page.html
```
<div id="mySwiper" class="swiper-container">
  <div class="swiper-wrapper">
    <div class="swiper-slide">Slide 1</div>
    <div class="swiper-slide">Slide 2</div>
    <div class="swiper-slide">Slide 3</div>
  </div>
  <div class="swiper-button-next"></div>
  <div class="swiper-button-prev"></div>
</div>
```
(aside: a non-``<ion-slides>`` ``.swiper-container`` doesn't swipe properly inside an ``<ion-content>`` — probably a separate issue there that needs to be addressed and may be related to #16579)

page.ts
```
import { default as Swiper } from 'swiper';

  // ...
  ngAfterViewInit() {
    new Swiper('#mySwiper', {
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev',
      },
    });
  }
```

As for changes that can be made to ``<ion-slides>``, a possibility is to add a property that accepts an alternative Swiper library to be used when ``ion-slides`` is init'ed.

i.e. in [slides.tsx](https://github.com/ionic-team/ionic/blob/v4.0.0-rc.0/core/src/components/slides/slides.tsx#L294), replace:
```
  const swiper = new Swiper(this.el, finalOptions);
```

with something like:
```
  @Prop() swiperlib: any;
  // ...
  const swiper = new (this.swiperlib || Swiper)(this.el, finalOptions);
```

With that additional property and after an ``npm i swiper`` in your project, in your page.html:
```
  <ion-slides [swiperlib]="mySwiperLib">
```

and in page.ts:
```
import { default as Swiper } from 'swiper';

export class Page {
  mySwiperLib = Swiper;
  // ...
```

Seems a tad ugly in my opinion though to have a property solely for passing a different version of the Swiper library. It would duplicate existing code as you've already pointed out. But doing so would allow you to use all the swiperOptions if you provide the proper library via the ``swiperlib`` property. Your project will have to have either a dependency on swiper or include a different custom built bundle.

**Comment by CFT-Chris** [comment]
I would like to submit a PR regardless, either exposing a ``swiperlib`` property as I described in my previous comment, or for my own sake, implementing parallax and navigation arrows (exposing ``parallax`` and ``navigation`` properties like we have for ``pager`` and ``scrollbar``, or maybe more simply looking inside the ``options`` property to check if parallax and navigation are enabled in SwiperOptions).

Just need input from ionic-team (perhaps @paulstelzer can change label to ask for input?) prior to doing any of this work.

**Comment by ivanmayes** [comment]
+1, as a user of Ionic long before swiper was implemented, it just seems silly to keep trying to stay up-to-date with the progress being made on that plugin.  Things like slideToLoop are critical for more complex slides and just aren't exposed or updated frequently enough to depend on ionSlides.

Exposure of the swiper instance would at least allow power user functionality of the current plugin, and would allow you to not have to spend as much time keeping up with the API.  Also, I'm pretty sure this was exposed in 3, so not sure why it became private.

**Comment by CFT-Chris** [comment]
FWIW, the solution I proposed on Dec 28, 2018 to expose a ``swiperlib`` property works fine for me and I have local changes that I can push towards a PR.

That may be the best solution going forward seeing how there is variable dependency among Ionic users on modules not included in the packaged Swiper bundle (see #16616, e.g.), and users wanting to use the latest Swiper code without having to wait on Ionic to update its bundle. It would resolve a bunch of Swiper issues at once.

I also have local changes for implementing just ``navigation`` and ``parallax`` properties alone which would resolve this issue and my particular issue (not logged) with missing navigation arrows. Just looking for feedback from @manucorporat or any of the Ionic team on how they want to proceed with Swiper integration before I submit any PRs.

**Comment by isnob46** [comment]
there is any update of this?
i really want to use full swiper api, mostly the the virtual slide feature,
There is any way to use ionic slides with virtual slides?

**Comment by markoivanovski** [comment]
Thanks for adding the `getSwiper()` method @liamdebeasi. My team and I are trying to use some of the other modules from Swiper but I found this line in the source code which indicates Ionic limits to AutoPlay, Pagination, Scrollbar and Zoom.

`import { Autoplay, Pagination, Scrollbar, Swiper, Zoom } from 'swiper/dist/js/swiper.esm';`
https://github.com/ionic-team/ionic/blob/master/core/src/components/slides/swiper/swiper.js

Is there or will there be away to configure the Slider to use additional modules? 

Thanks!
